### PR TITLE
Allow for processing multiple workflows simultaneously

### LINKED ISSUE #26: Launching multiple workflows simultaneously
Pytest-workflow now runs the workflows one at a time. This can be very slow.
If more compute power is available (for example in a cluster environment), there should be an option to increase the number of workflows spwaned.

### PR DESCRIPTION
**Pull request by rhpvorderman**
Using the `--wt` or `--workflow-threads` flag, the number of workflows executed simultaneously can be set (default=1). 

This required some refactoring of the code. I hope the motivation for certain designs is clearly explained by the comments in the code. If not, please add review comments on this.

This fixes #26 .

### Checklist
- [x] Pull request details were added to HISTORY.rst

### CODE REVIEW

**Comment by rhpvorderman** [comment]
codacy has a false positive on an unused import.

File: src/pytest_workflow/workflow.py [hunk 1]
```diff
@@ -17,54 +17,171 @@
...
+        return self._stderr.decode().splitlines()
 
 
 def bytes_to_file(bytestring: bytes, output_file: Path) -> Path:
     with output_file.open('wb') as file_handler:
         file_handler.write(bytestring)
     return output_file
+
+
+class WorkflowQueue(queue.Queue):
```
Comment (DavyCats) [inline src/pytest_workflow/workflow.py]: It might be preferable to have this in a separate file.

Comment (rhpvorderman) [reply src/pytest_workflow/workflow.py]: The file is less than 200 lines including license. What would be the benefits of splitting it up?

File: src/pytest_workflow/content_tests.py [hunk 2]
```diff
@@ -88,33 +90,61 @@ def file_to_string_generator(filepath: Path) -> Iterable[str]:
 
 class ContentTestCollector(pytest.Collector):
     def __init__(self, name: str, parent: pytest.Collector,
-                 content: Iterable[str], content_test: ContentTest):
+                 content_generator: Callable[[], Iterable[str]],
```
Comment (sndrtj) [inline src/pytest_workflow/content_tests.py]: The `content_generator` takes no arguments? i.e, its output is always the same?

File: src/pytest_workflow/content_tests.py [hunk 3]
```diff
@@ -124,24 +154,27 @@ def collect(self):
...
         contain = "contains" if should_contain else "does not contain"
         name = "{0} '{1}'".format(contain, string)
         super().__init__(name, parent=parent)
         self.should_contain = should_contain
         self.string = string
-        self.contains = contains
 
     def runtest(self):
-        assert self.contains == self.should_contain
+        self.parent.thread.join()
```
Comment (sndrtj) [inline src/pytest_workflow/content_tests.py]: Perhaps a little comment to indicate why this join is performed would be nice.

File: src/pytest_workflow/file_tests.py [hunk 4]
```diff
@@ -105,20 +118,23 @@ def repr_failure(self, excinfo):
...
         """
         name = "md5sum"
         super().__init__(name, parent)
         self.filepath = filepath
         self.expected_md5sum = md5sum
         self.observed_md5sum = None
+        self.workflow = workflow
 
     def runtest(self):
+        self.workflow.wait()
```
Comment (sndrtj) [inline src/pytest_workflow/file_tests.py]: Why `wait` here while `ContetentTestItem` uses `join`? Same question for the `FileExists` class.

File: src/pytest_workflow/plugin.py [hunk 5]
```diff
@@ -52,6 +58,28 @@ def pytest_collect_file(path, parent):
...
+    # We need to add a workflow queue to some central variable. Instead of
+    # using a global variable we add a value to the config.
+    # Configuring was copied from _pytest/tmpdir.py. There they also add
+    # objects to the config.
+    # pylint: disable=protected-access
+    # protected access is needed here to do it the pytest way: ugly.
+    monkeypatch = MonkeyPatch()
+    config._cleanup.append(monkeypatch.undo)
+    workflow_queue = WorkflowQueue()
+    monkeypatch.setattr(config, "workflow_queue", workflow_queue,
```
Comment (sndrtj) [inline src/pytest_workflow/plugin.py]: I know this is pytest's internals showing its guts, but man, this should make everyone weep. 
Does python's builtin `setattr` not work here? I.e. simply `setattr(config, 'workflow_queue', workflow_queue)` ?

File: src/pytest_workflow/plugin.py [hunk 6]
```diff
@@ -117,18 +147,31 @@ def run_workflow(self):
         # Create a workflow and make sure it runs in the tempdir
         workflow = Workflow(self.workflow_test.command, tempdir)
 
-        print("run '{name}' with command '{command}' in '{dir}'".format(
+        # Add an extra print. As it looks better in the pytest output.
+        print()
+        print("'{name}' with command '{command}' in '{dir}' is queued.".format(
```
Comment (sndrtj) [inline src/pytest_workflow/plugin.py]: You can also start the second print statement with a newline. That looks a bit cleaner, i think. 
I.e. `print("\n'{name}' with command ......)`

File: src/pytest_workflow/workflow.py [hunk 7]
```diff
@@ -17,54 +17,171 @@
...
+                        " last {0} seconds".format(self.wait_timeout_secs))
+                time.sleep(self.wait_interval_secs)
+                self.wait_counter += 1
+
+            # Wait for process to finish with _popen.communicate().
+            # _popen.wait() will block with stdout=pipe and stderr=pipe
+            if (self._popen.returncode is None and
+                    self._stderr is None and
+                    self._stdout is None):
+                self._stdout, self._stderr = self._popen.communicate()
```
Comment (sndrtj) [inline src/pytest_workflow/workflow.py]: so `communicate()` blocks until the command completes?

File: src/pytest_workflow/workflow.py [hunk 8]
```diff
@@ -17,54 +17,171 @@
...
+                #    wait only once for the workflow to start. All consecutive
+                #    wait commands will fail instantly. Having all of these
+                #    wait as well would be a waste of time.
+                if self.wait_timeout_secs is not None and self.wait_counter > (
+                        self.wait_timeout_secs / self.wait_interval_secs):
+                    raise ValueError(
+                        "Waiting on a workflow that has not started within the"
+                        " last {0} seconds".format(self.wait_timeout_secs))
+                time.sleep(self.wait_interval_secs)
+                self.wait_counter += 1
```
Comment (sndrtj) [inline src/pytest_workflow/workflow.py]: wouldn't it be easier to simply count on the number of seconds? i.e

```
if self.wait_time_seconds > self.max_wait_seconds:
    raise ValueError
time.sleep(self.wait_interval_secs)
self.wait_time_secs += self.wait_internval_secs
```

File: src/pytest_workflow/workflow.py [hunk 9]
```diff
@@ -17,54 +17,171 @@
...
+        self.wait()
+        return self._stderr  # for testing log
 
     @property
     def exit_code(self) -> int:
-        return self._proc_out.returncode
+        self.wait()
+        return self._popen.returncode
+
+    def stdout_lines(self):
```
Comment (sndrtj) [inline src/pytest_workflow/workflow.py]: returntype?

File: src/pytest_workflow/workflow.py [hunk 10]
```diff
@@ -17,54 +17,171 @@
...
+        self.wait()
+        return self._stdout.decode().splitlines()
+
+    def stderr_lines(self):
```
Comment (sndrtj) [inline src/pytest_workflow/workflow.py]: returntype?

File: src/pytest_workflow/workflow.py [hunk 11]
```diff
@@ -17,54 +17,171 @@
...
 import subprocess  # nosec: security implications have been considered
+import threading
+import time
 from pathlib import Path
+from typing import Optional  # noqa: F401  # used for typing
 
 
 class Workflow(object):
-
+    # pylint: disable=too-many-instance-attributes
```
Comment (sndrtj) [inline src/pytest_workflow/workflow.py]: Meta-comment: Nearly all methods/properties do `self.wait()`. Perhaps using a decorator makes it it a bit clearer as to why this happens, and allows to completely refactor the code without touching the multithreading code.

File: src/pytest_workflow/workflow.py (hunk 1 above)
Comment (sndrtj) [reply src/pytest_workflow/workflow.py]: Agreed with @rhpvorderman here.

File: src/pytest_workflow/file_tests.py (hunk 4 above)
Comment (rhpvorderman) [reply src/pytest_workflow/file_tests.py]: `wait` is a function of `Workflow`. `join()` is a function of thread.

File: src/pytest_workflow/workflow.py (hunk 8 above)
Comment (rhpvorderman) [reply src/pytest_workflow/workflow.py]: But that's floating point calculation. We should use integers for premature optimization overengineering concerns!!!1!1!!1!

File: src/pytest_workflow/workflow.py (hunk 11 above)
Comment (rhpvorderman) [reply src/pytest_workflow/workflow.py]: Added this to the to-do list #32

**Pull request merged**